### PR TITLE
Update old Hourly/Daily data for ApiDataProvider

### LINKED ISSUE #3001: Daily and hourly data zips do not re-download and stay stale
#### Expected Behavior
LEAN re-downloads daily/hourly zips when it encounters missing days.

#### Actual Behavior
Right now, if I have a daily/hourly zip downloaded as of a week ago, my backtests will run successfully to today but the price data used will be the last close in the zip, rendering the backtests useless. LEAN doesn't understand the need to re-download the zips.

#### Potential Solution
Any solution I can think of would fail in the case of weekends/holidays (i.e. it would trigger a useless zip re-download on such days). There should be a way to know if a date should have data and, thus, if it doesn't, to re-download the zip.

#### Reproducing the Problem
Download a daily/hourly zip as of today. Come back a few days later and run a daily/hourly backtest that extends past today's date. Log output of price to observe staleness.

#### System Information
Windows x64

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Have ApiDataProvider update stale hourly/daily data before returning a filestream.

#### Related Issue

Closes #3001 

#### Motivation and Context

Hourly/Daily data is stored in a single file so using ApiDataProvider with Lean leads to a lot of stale data. This PR seeks to add an auto update feature for the data on a configurable period.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

Refactor has been tested, but I am unable to checkout daily data in web data repo so I am unable to test the replacement.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Engine/DataFeeds/ApiDataProvider.cs [hunk 1]
```diff
@@ -51,40 +52,70 @@ public ApiDataProvider()
...
+        /// Determine if the file is out of date based on configuration and needs to be updated
+        /// </summary>
+        /// <param name="resolution">Data resolution</param>
+        /// <param name="filepath">Path to the file</param>
+        /// <returns>True if the file is out of date</returns>
+        /// <remarks>Files are only "out of date" for Hourly/Daily data because this data is stored all in one file</remarks>
+        private bool IsOutOfDate(Resolution resolution, string filepath)
+        {
+            return resolution >= Resolution.Hour &&
+                (DateTime.Now - TimeSpan.FromDays(_downloadPeriod)) > File.GetLastWriteTime(filepath);
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ApiDataProvider.cs:85]: ah nice! 👍 think we should be able to make couple of simple unit test cases for this guy

File: Tests/Engine/DataProviders/ApiDataProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        [TestCase(Resolution.Minute, 6, false)]
+        [TestCase(Resolution.Second, 6, false)]
+        [TestCase(Resolution.Tick, 6, false)]
+        public void OutOfDateTest(Resolution resolution, int days, bool expected)
+        {
+            var path = "./testfile.txt";
+            var time = DateTime.Now - TimeSpan.FromDays(days);
+
+            File.Create(path).Close();
+            File.SetLastWriteTime(path, time);
```
Comment (Martin-Molinero) [inline Tests/Engine/DataProviders/ApiDataProviderTests.cs:48]: Cool! 💯

File: Engine/DataFeeds/ApiDataProvider.cs [hunk 3]
```diff
@@ -51,40 +52,70 @@ public ApiDataProvider()
...
+        }
+
+        /// <summary>
+        /// Determine if the file is out of date based on configuration and needs to be updated
+        /// </summary>
+        /// <param name="resolution">Data resolution</param>
+        /// <param name="filepath">Path to the file</param>
+        /// <returns>True if the file is out of date</returns>
+        /// <remarks>Files are only "out of date" for Hourly/Daily data because this data is stored all in one file</remarks>
+        public bool IsOutOfDate(Resolution resolution, string filepath)
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ApiDataProvider.cs:82]: nit - maybe we can make this method static as well as `_downloadPeriod`

**Review by Martin-Molinero — APPROVED** [review]
Another nice one! leaving a nit comment

**Review by jaredbroad — APPROVED** [review]

**Pull request merged**